test/devicecontroller/DefaultDeviceController.test.ts

```diff
@@ -1037,25 +1035,300 @@ describe('DefaultDeviceController', () => {
+      deviceController.releaseMediaStream(stream);
+      await new Promise(resolve => setTimeout(resolve, 200));
+      await deviceController.chooseVideoInputDevice(null);
+
+      expect(called).to.be.true;
+      await device.stop();
+    });
+
+    it('can choose media stream as device then choose VideoTransformDevice without audio video controller', done => {
+      domMockBehavior.getUserMediaSucceeds = true;
+      const mockVideoStream = new MediaStream();
+      // @ts-ignore
+      mockVideoStream.active = true;
+      // @ts-ignore
+      mockVideoStream.id = 'sample';
+      // @ts-ignore
+      const mockVideoTrack = new MediaStreamTrack('test', 'video');
+      mockVideoStream.addTrack(mockVideoTrack);
+      domMockBehavior.createElementCaptureStream = mockVideoStream;
+      deviceController.chooseVideoInputDevice(mockVideoStream).then(() => {});
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, mockVideoStream, [processor]);
+      deviceController.chooseVideoInputDevice(device).then(() => {});
+      new TimeoutScheduler(400).start(() => {
+        deviceController.chooseVideoInputDevice(null);
+      });
+
+      new TimeoutScheduler(450).start(() => {
+        device.stop();
+        done();
+      });
+    });
+
+    it('can choose MediaStream as device then choose VideoTransformDevice when startLocalVideoTile is called', done => {
+      class TestAudioVideoController extends NoOpAudioVideoController {
+        async replaceLocalVideo(): Promise<void> {
+          return;
         }
-        intrinsicDevice(): Promise<Device> {
-          throw new Error('Method not implemented.');
+      }
+      domMockBehavior.getUserMediaSucceeds = true;
+      const mockVideoStream = new MediaStream();
+      // @ts-ignore
+      mockVideoStream.active = true;
+      // @ts-ignore
+      mockVideoStream.id = 'sample';
+      // @ts-ignore
+      const mockVideoTrack = new MediaStreamTrack('test', 'video');
+      // @ts-ignore
+      mockVideoTrack.kind = 'video';
+      mockVideoStream.addTrack(mockVideoTrack);
+      domMockBehavior.createElementCaptureStream = mockVideoStream;
+      const avController = new TestAudioVideoController();
+      const spy = sinon.spy(avController, 'replaceLocalVideo');
+
+      deviceController.bindToAudioVideoController(avController);
+
+      avController.videoTileController.startLocalVideoTile();
+      deviceController.chooseVideoInputDevice(mockVideoStream).then(() => {});
+      const processor = new NoOpVideoFrameProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      deviceController.chooseVideoInputDevice(device).then(() => {});
+      new TimeoutScheduler(400).start(() => {
+        deviceController.chooseVideoInputDevice(null);
+      });
+
+      new TimeoutScheduler(450).start(() => {
+        device.stop();
+        expect(spy.called).to.equal(true);
+        done();
```

Comment:
Generally speaking I want to ban the use of `TimeoutScheduler` in tests unless you are specifically assessing a time-based piece of code.